File: Java/src/com/leetcode/Easy/LastStoneWeight.java
package com.leetcode.Easy;

import java.util.*;

// 1046. Last Stone Weight
public class LastStoneWeight {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for (int i : stones)
            pq.offer(i);

        while (pq.size() > 1) {
            int max1 = pq.poll();
            int max2 = pq.poll();

            if (max1 - max2 > 0)
                pq.offer(max1 - max2);
        }

        return pq.size() > 0 ? pq.poll() : 0;
    }
}

File: JavaScript/Easy/LastStoneWeight.js
/**
 * 1046. Last Stone Weight
 * @param {number[]} stones
 * @return {number}
 */
var lastStoneWeight = function (stones) {
  while (stones.length > 1) {
    stones.sort((a, b) => b - a);
    let firstMax = stones[0];
    let secondMax = stones[1];
    stones.splice(0, 2);
    if (firstMax !== secondMax) {
      stones.push(firstMax - secondMax);
    }
  }
  
  return stones.length === 0 ? 0 : stones[0];
};
